cmd/web-repo-proxy/main.go

```diff
@@ -0,0 +1,323 @@
+	return filepath.Join(repositoriesRoot, string(r))
+}
+
+// urlPath returns the URL path to use when cloning this repository over http.
+func (r repository) urlPath() string {
+	return repositoryRequestPrefix + string(r)
+}
+
+// goodUntil checks the repository's directory on disk and returns the time
+// it will expire, or an error if the repository couldn't be read.
+func (r repository) goodUntil() (time.Time, error) {
+	path := r.filePath()
+	info, err := os.Lstat(path)
+	if err != nil {
+		return time.Time{}, err
+	}
+	return info.ModTime().Add(repositoryTTL), nil
+}
+
+func createRandomKey() string {
+	keyIndex := rand.Int31()
+	return strconv.FormatInt(int64(keyIndex), 16)
+}
+
+// createNewRepository creates a new repository with the given name and data.
+// Returns a repository string if successful, otherwise returns error.
+func createNewRepository(name string, data map[string]string) (repository, error) {
+	// Create a temporary directory to initialize with the repository data.
+	tempRoot, err := ioutil.TempDir("", "web-repo-proxy")
+	if err != nil {
+		return "", err
+	}
+	defer func() {
+		// Make sure we remove the temporary directory, in case there was an
+		// error before it was moved to its final location  (otherwise this is a
+		// no-op).
+		os.RemoveAll(tempRoot)
+	}()
+
+	// Write out the given data to the file system.
+	for filename, content := range data {
+		filePath := filepath.Join(tempRoot, filename)
+		if err = ioutil.WriteFile(filePath, []byte(content), 0644); err != nil {
+			return "", err
+		}
+	}
+
+	// Set up the git repository.
+	gitCommands := [][]string{
+		[]string{"init"},
+		[]string{"add", "."},
+		[]string{"commit", "-m", "Initial commit"},
+		// update-server-info is needed to serve repositories over static http.
+		[]string{"update-server-info"},
+	}
+	for _, command := range gitCommands {
+		cmd := exec.Command("git", command...)
+		cmd.Dir = tempRoot
+		if err = cmd.Run(); err != nil {
+			return "", err
+		}
+	}
+
+	// Choose the real repository path and move it there.
+	for i := 0; i < 3; i++ {
+		// Retry up to 3 times to avoid ephemeral errors and name collisions.
+		repository := repository(name + "-" + createRandomKey())
+		repositoryRoot := repository.filePath()
+		if err = os.Rename(tempRoot, repositoryRoot); err == nil {
+			return repository, nil
+		}
+	}
+	// If we still failed after 3 tries, return the most recent error.
+	return "", err
+}
+
+// Check whether any repositories have expired, and delete them if so.
+func deleteExpiredRepositories() {
+	curTime := time.Now()
+	files, err := ioutil.ReadDir(repositoriesRoot)
+	if err != nil {
+		log15.Error("Couldn't read repositories directory", "error", err)
+		return
+	}
+	for _, file := range files {
+		if file.ModTime().Add(repositoryTTL).Before(curTime) {
+			// This repository has expired.
+			tempDir, err := ioutil.TempDir("", "web-repo-proxy")
+			if err != nil {
+				log15.Error("Couldn't create temporary directory", "error", err)
```

Comment:
To keep this log consistent with the other log at the end of the function:

```suggestion
				log15.Error("Couldn't create temporary directory", "tempDir", tempDir, "error", err)
```